src/poetry/core/version/markers.py

```diff
@@ -259,21 +260,28 @@ def value(self) -> str:
 
     def intersect(self, other: MarkerTypes) -> MarkerTypes:
         if isinstance(other, SingleMarker):
-            if other.name != self.name:
-                return MultiMarker(self, other)
+            if self.name == other.name or (
+                self.name in PYTHON_VERSION_MARKERS
+                and other.name in PYTHON_VERSION_MARKERS
+            ):
+                if self.constraint == other.constraint:
+                    return self
 
-            if self == other:
-                return self
+                if self._operator in {"in", "not in"} or other.operator in {
+                    "in",
+                    "not in",
+                }:
+                    return MultiMarker.of(self, other)
 
-            if self._operator in {"in", "not in"} or other.operator in {"in", "not in"}:
-                return MultiMarker.of(self, other)
+                new_constraint = self._constraint.intersect(other.constraint)
+                if new_constraint.is_empty():
+                    return EmptyMarker()
 
-            new_constraint = self._constraint.intersect(other.constraint)
-            if new_constraint.is_empty():
-                return EmptyMarker()
+                if new_constraint == self.constraint:
+                    return self
 
-            if new_constraint == self._constraint or new_constraint == other.constraint:
-                return SingleMarker(self._name, new_constraint)
+                if new_constraint == other.constraint:
+                    return other
```

Comment:
I've not checked in full detail, but I have the feeling that this part of code can be removed for the most part after your other PRs (#282 and #293) have been merged because it's a duplication of the code in `MultiMarker.of()`. Thus, I'd propose to wait for them, rebase and check if my feeling is right.

Reply:
I think you're right and this section can become just

```python
if isinstance(other, SingleMarker):
    return MultiMarker.of(self, other)
```

So I agree it makes sense to do others first and perhaps this can be simplified